### PR TITLE
APIv4 - Remove gatekeeper permission check for AJAX endpoint (with tests)

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Historically the AJAX api has required either 'access CiviCRM' or 'access AJAX API' as a broad hedge against security vulnerabilities, but this interferes with legitimate use-cases for anonymous users to make ajax requests. Since http requests are allowed by anon, and many Civi pages now contain ajax elements, this PR replaces that broad hedge with unit tests for AJAX security.

Before
----------------------------------------
- APIv4 AJAX requires either 'access CiviCRM' or 'access AJAX API' permission, even though some APIs like `Contact.get`, `SearchDisplay.run` etc. do not require any permissions (instead they check ACLs).
- No unit tests for APIv4 AJAX endpoint.
- When making multiple calls in one request, the entire request will fail if one call fails permission checks.

After
----------------------------------------
- Gatekeeper check removed for AJAX APIv4 (nothing changed for v3).
- `hook_ alterApiRoutePermissions` removed as no longer needed. 
- Unit tests added for APIv4 AJAX endpoint.
- Multiple calls in a single request will not cause each other to fail.

Comments
----------------------------------------
I hit this when working on SearchKit and needing to mix multiple AJAX calls in a single request. For anonymous users it failed miserably without this PR. Now it works :)

### CODE REVIEW

File: CRM/Utils/Hook.php [hunk 1]
```diff
@@ -3081,41 +3081,6 @@ public static function alterCustomFieldDisplayValue(&$displayValue, $value, $ent
...
-   * - if it is found that "access AJAX API" guard is not needed for APIv4.
-   * - if the policy is moved into metadata.
-   *
-   * @param array|string $permissions
-   * @param string $entity
-   * @param string $action
-   *
-   * @return mixed
-   */
-  public static function alterApiRoutePermissions(&$permissions, $entity, $action) {
```
Comment (totten) [inline CRM/Utils/Hook.php:3109]: I did a search of `universe` and only found one other consumer of this hook (cc @MegaphoneJon):

https://github.com/MegaphoneJon/com.megaphonetech.monitoring/blob/86e0776aafb46e3181049551f8bb700f1aa0cf54/monitoring.php#L78

However, I think this is OK -- the route-level guard goes away, but the API-level guard is still there. And `hook_alterAPIPermissions` provides a similar adjustment for the API-level guard.

**Comment by totten** [comment]
I tend to agree that `access AJAX API` feels a bit unnecessary and looks like an old hedge - and that it's continuing existence just invites misconfiguration. But since this is a bit speculative, and since it's a fairly central permission, I think we should ping a bigger list in case there was angle we don't see. (*I'll ping the sec-team.*)

FWIW, there are several ext's in `universe` that make reference to `access AJAX API`. So it's probably good that this PR leaves the declaration (*i.e. the permission nominally exists - it's just not needed by any of the original core things*).

I would probably vote for this to go into 5.64-dev (*after the freeze for 5.63, which is coming soon*).

**Comment by kcristiano** [comment]
@colemanw Thank you for that clarification.   I had started testing on this with some extensions and WP plugins that rely on `access_ajax_api`  

Based on your comment I can limit the testing to API4 only?   If this gets merged is the plan to remove from API3 as well?   I want to be sure I run through my testing properly.

**Comment by colemanw** [comment]
@kcristiano thanks for testing!
I did not have any plans to alter APIv3, as I think there is general agreement that v3 is frozen except for regression fixes.

**Comment by MegaphoneJon** [comment]
Thanks for the ping.  Not surprising that there aren't many consumers since this hook is undocumented!  I don't have a comment on the larger implications, but I'm fine with this going away in the case of my extension.

**Comment by colemanw** [comment]
Retest this please

**Comment by colemanw** [comment]
@artfulrobot I don't get it. I just rebased this PR to the tip of `master` and the newly-built D7 demo site seems to work fine.
Can you try clearing caches and rebuilding menus after applying the patch? This PR does alter a menu route permission so it could be a caching issue.

**Comment by kcristiano** [comment]
@colemanw 

I've applied this to a site that is heavy on front end Forms with varying permissions.   

 `r-run`  passed and worked as expected.   

I agree with @totten that its merge ready and I'd rather seen it in sooner than later.

**Pull request merged**